src/lib/type_check.ml

```diff
@@ -3594,48 +3595,60 @@ and infer_exp env (E_aux (exp_aux, (l, uannot)) as exp) =
-            end
-        | None -> begin
-            match typ_of then_branch' with
-            | Typ_aux (Typ_app (f, [_]), _) when string_of_id f = "atom_bool" ->
-                let else_branch' =
-                  crule check_exp
-                    (add_opt_constraint l "else branch" (Option.map nc_not (assert_constraint env false cond')) env)
-                    else_branch bool_typ
-                in
-                annot_exp (E_if (cond', then_branch', else_branch')) bool_typ
-            | _ ->
-                let else_branch' =
-                  crule check_exp
-                    (add_opt_constraint l "else branch" (Option.map nc_not (assert_constraint env false cond')) env)
-                    else_branch (typ_of then_branch')
-                in
-                annot_exp (E_if (cond', then_branch', else_branch')) (typ_of then_branch')
-          end
-      end
+
+      (* When one branch's type is inferred, check the other branch's type against it. *)
+      let one_branch_inferred inferred_branch other_branch other_env =
+        (* If the type of the inferred branch is a function call to `atom_bool` treat it as bool. *)
+        let inferred_typ = if is_atom_bool (typ_of inferred_branch) then bool_typ else typ_of inferred_branch in
+        (* Check the other branch matches the type. *)
+        let other_branch' = crule check_exp other_env other_branch inferred_typ in
+        (other_branch', inferred_typ)
+      in
+
+      match (branch_typ then_branch then_env, branch_typ else_branch else_env) with
+      (* Both branches are simple numeric types. *)
+      | Left (then_branch', Some then_sn), Left (else_branch', Some else_sn) ->
+          let cond_constraint = destruct_atom_bool env (typ_of cond') in
+          let typ = typ_of_simple_numeric (union_simple_numeric cond_constraint then_sn else_sn) in
+          annot_exp (E_if (cond', then_branch', else_branch')) typ
+      (* One branch is a simple numeric type but the other isn't. *)
+      | Left (_, Some _), _ -> typ_error l ("Could not infer type of " ^ string_of_exp else_branch)
+      | _, Left (_, Some _) -> typ_error l ("Could not infer type of " ^ string_of_exp then_branch)
+      (* Neither branch is a simple numeric type, but we inferred the `then` branch. *)
+      | Left (then_branch', None), _ ->
+          let other_branch, inferred_typ = one_branch_inferred then_branch' else_branch else_env in
+          annot_exp (E_if (cond', then_branch', other_branch)) inferred_typ
+      (* Neither branch is a simple numeric type, but we inferred the `else` branch (but not the `then` branch). *)
+      | _, Left (else_branch', None) ->
+          let other_branch, inferred_typ = one_branch_inferred else_branch' then_branch then_env in
+          annot_exp (E_if (cond', other_branch, else_branch')) inferred_typ
+      (* We couldn't infer the type of either branch; raise the error for the `then` branch. *)
+      | Right err, _ -> raise err
```

Comment:
We should display both errors here, but that would require a new `type_error` member with custom display logic. I can add that later if you want.

Reply:
Yeah I don't think it's super critical though; I don't really expect compilers to find all errors after the first one anyway.